### DISCUSSION TITLE
Feature request: Link to download copy of notebook

### DISCUSSION THREAD

**Discussion by andrewcstewart** [quarto-dev/quarto-cli#2432] [Feature Requests]
It would be great if Quarto could allow you to generate a link in the navbar (or wherever) to download a copy of the notebook you are viewing.  

Other frameworks like Jupyterbook implement examples of this feature.  Or,  a picture is worth a thousand words, as they say...
<img width="1087" alt="image" src="https://user-images.githubusercontent.com/1420654/190014713-e2639b2b-5555-4e90-9677-29c3b2c28583.png">

**Comment by cderv** [comment]
Quarto has a `code-tools` option you can activate to show the source of the document 
https://quarto.org/docs/output-formats/html-code.html#code-tools

You can also add download button in the sidebar by setting `download` options under `book`. See Sidebar tools (https://quarto.org/docs/books/book-output.html#sidebar-tools) and all the `book` key options (https://quarto.org/docs/reference/projects/books.html#meta-project-type-upper)

They would not be at the same place, and that would a feature request to allow to place the source view / download inside the sidebar tool 🤔

**Reply by jav-ed** [reply]
Maybe you want to have a look at the [downloadthis extension](https://github.com/shafayetShafee/downloadthis). It also allows to download .ipynb files. However, I do not know if the download button can be inserted inside the sidebar or any other place than the writing space.

**Reply by andrewcstewart** [reply]
Hi @jav-ed - Cool extension!  So could I probably do something like set `keep-ipynb: true` and then include `{{< downloadthis myfile.ipynb >}}` in the page?

**Reply by jav-ed** [reply]
I am not using quarto anymore. I was comparing quarto against jupyterbook. Furthermore, I only spent a little bit of time with quarto, thus I do not know about keep-ipynb: true, but  {{< downloadthis myfile.ipynb >}} was tested and it worked. 

By the way, just because I use jupyter book that does not mean, you should too. Quarto is an amazing work, with a huge amount of features. It was very hard for me to decide between quarto and jupyter book

**Comment by dragonstyle** [comment]
If you specify `ipynb` as an output format in your Quarto document using Quarto 1.3 (currently pre-release), we will produce a rendered version of the Notebook (in which things like cross refs and citations have been resolved) and make it available in a menu in the TOC. 

(more here https://quarto.org/docs/prerelease/1.3/multi-format.html)

**Reply by TS-CUBED** [reply]
Love this feature.

However, it doesn't work on book level. Will this be possible on book level as well? E.g. one notebook per page?

**Reply by dragonstyle** [reply]
It is currently walled off from this case, but this is something we would like to support ( as well as potentially marking up a page and 'snipping' notebooks of computational cells out of the page as supporting files). We won't get to it in 1.3, but I will open an enhancements request to track it for 1.4.

**Reply by TS-CUBED** [reply]
I have a workaround for now based on 

`keep-ipynb: true`

which keeps a Jupyter notebook for each qmd that has code cells in it, and  the `downloadthis` extension, which embeds download buttons in the page.

The created notebooks are a bit broken, though, with respect to cross-references and such.

Also been playing around with the `noteout` pandoc filter which works, and allows the snipping feature you mention (the only annoyance here is the "Beginning of `name` notebook" text in the main text of the document. 

Will need to work to combine the two somehow (or wait for v1.4) ;-)

**Comment by mcanouil** [comment]
I believe the "embedding Jupyter Notebooks" features address at least partially this (https://quarto.org/docs/authoring/notebook-embed.html).
Also closing in favour of #4674.
